### PR TITLE
Deployer : Skip batches when nothing has changed

### PR DESCRIPTION
**Pull request by ilausuch**
By using skip_batch_for_node? we can avoid running chef on nodes that do
not change role and if there is no attribute change for the barclamp.

Does not skip nodes that have changed roles, all nodes if
attributes have changed and when deploying the barclamp for the first
time.

Please help potential reviewers to understand this pull request and speed
up the process by writing a meaningful pull request message.

### CODE REVIEW

**Comment by ilausuch** [comment]
Tested on 8.ha

File: crowbar_framework/app/models/deployer_service.rb [hunk 1]
```diff
@@ -138,4 +138,21 @@ def transition(inst, name, state)
...
+  # try to know if we can skip a node from running chef-client
+  def skip_unchanged_node?(node_name, old_role, new_role)
+    # if old_role is nil, then we are applying the barclamp for the first time
+    return false if old_role.nil?
+
+    # if the node changed roles, then we need to apply
+    return false if node_changed_roles?(node_name, old_role, new_role)
+
+    # if attributes have changed, we need to run
+    return false if node_changed_attributes?(node_name, old_role, new_role)
```
Comment (vuntz) [inline crowbar_framework/app/models/deployer_service.rb]: That's actually not true for the deployer. Can you remove the code and add an explicit comment? (the attributes are only used by the rails app)

File: crowbar_framework/app/models/deployer_service.rb [hunk 2]
```diff
@@ -138,4 +138,20 @@ def transition(inst, name, state)
...
+
+  # try to know if we can skip a node from running chef-client
+  def skip_unchanged_node?(node_name, old_role, new_role)
+    # if old_role is nil, then we are applying the barclamp for the first time
+    return false if old_role.nil?
+
+    # if the node changed roles, then we need to apply
+    return false if node_changed_roles?(node_name, old_role, new_role)
+
+    # There is not need to check attributes because they aren't necesarry for the current barclamp
```
Comment (vuntz) [inline crowbar_framework/app/models/deployer_service.rb]: Typo: "necessary"

Also, I would rephrase to "they are not used in cookbooks"; they are actually used in the barclamp (in the rails application).

**Comment by vuntz** [comment]
Hrm, something is wrong, there are 4 commits which shouldn't be here.

**Comment by vuntz** [comment]
> Hrm, something is wrong, there are 4 commits which shouldn't be here.

I fixed this.
